File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        const arr = []
        let curr = head;
        while(curr !== null){
            arr.push(curr);
            curr = curr.next
        }
        for(let i = arr.length-1;i>=0;i--){
            if(i === 0){
                arr[i].next = null;
                continue;
            }
            arr[i].next = arr[i-1];
        }
        if(arr.length > 0)
        head = arr[arr.length-1]
        return head
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> q = new PriorityQueue<>((a, b) -> a.val - b.val);
        for(ListNode l: lists){
            q.offer(l);
        }
        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;
        while(!q.isEmpty()){
            ListNode minNode = q.poll();
            curr.next =minNode;
            curr = curr.next;

            if(minNode.next != null){
                q.offer(minNode.next);
            }
        }
        return dummy.next;
    }
}
